lib/linked-list.js

```diff
@@ -125,42 +200,118 @@ class LinkedList {
     /*
     method to reverse the singly linked list
     note: the nodes should be moved and not just the values in the nodes
-    Time Complexity: ?
-    Space Complexity: ?
+    Time Complexity: O(n)
+    Space Complexity: O(1)
     */
+   // 2 -> 3 -> 4 -> 5
     reverse() {
-        throw new Error("This method hasn't been implemented yet!");
+        if (!this.#head) {
+            return;
+        };
+        
+        let previous = null;
+        let current = this.#head;
+        let next = null;
+
+        while (current) {
+            next = current.next;
+            current.next = previous;
+            previous = current;     
+            current = next;     
+        }
+        this.#head = previous;
     }
 
     // Advanced Exercises
     /*
     returns the value at the middle element in the singly linked list
-    Time Complexity: ?
-    Space Complexity: ?
+    Time Complexity: O(n)
+    Space Complexity: O(1)
      */
     findMiddleValue() {
-        throw new Error("This method hasn't been implemented yet!");
+        let a = 0;
+        let current = this.#head;
+
+        while (current) {
+            a ++;
+            current = current.next;
+        };
+
+        let b = 0;
+        let middle = 0;
+
+        if (a % 2) {
+            middle = Math.floor(a/2);
+        } else {
+            middle = a/2;
+        };
+        
+        let current2 = this.#head;
+
+        while (current2) {
+            if (b === middle) {
+                return current2.value;
+            } 
+            b++;
+            current2 = current2.next;
+        };
     }
 
     /*
     find the nth node from the end and return its value
     assume indexing starts at 0 while counting to n
-    Time Complexity: ?
-    Space Complexity: ?
+    Time Complexity: O(n)
+    Space Complexity: O(1)
     */
     findNthFromEnd(n) {
-        throw new Error("This method hasn't been implemented yet!");
+        let a = 0;
+        let current = this.#head;
+
+        while (current) {
+            a ++;
+            current = current.next;
+        };
+
+        let b = 0;
+        let index = a - n
+        let current2 = this.#head
+
+        while (current2) {
+            if (b === index) {
+                return current2.value;
+            } 
+            b++;
+            current2 = current2.next;
+        };
+
+        return null;
     }
 
     /*
     checks if the linked list has a cycle. A cycle exists if any node in the
     linked list links to a node already visited.
     returns true if a cycle is found, false otherwise.
-    Time Complexity: ?
-    Space Complexity: ?
+    Time Complexity: O(n^2)
+    Space Complexity: O(n)
     */
     hasCycle() {
```

Comment:
👍  This works, but can you think of a way to do this in O(n) time complexity and O(1) space?